enrollment.md

```diff
@@ -0,0 +1,153 @@
+
+Each block produced by the enrollment network has a block height (a monatonic counter enforced by consensus), the corresponding AppHash, and a consensus timestamp. By signing a manifest that includes `(AppHash, height, timestamp)`, the manifest signer demonstrates that the manifest could not have existed before that block was finalized.
+
+Previous proofs (older AppHash values) can be reused in manifests, allowing for backdating. This is acceptable because the security goal is to prevent future-dating, wherein a manifest would be valid longer than it should be.
+
+The chain itself does not enforce expiry; clients check the expiry using the consensus timestamp.
+
+The consensus data will be published to a CDN with the following structure:
+
+```json
+{
+  "height": 12345,
+  "app_hash": "abc123def456...",
+  "timestamp": 1640995200,
+  "block_time": "2025-05-01T00:00:00Z"
+}
+```
+
+This allows clients to verify the `AppHash` against the chain and check manifest expiry using the consensus timestamp.
+
+## Enrollment
+
+Chain parameters:
+- `MAX_NUMBER_SUBDOMAINS`: the maximum number of subdomains per domain to rate-limit abuse.
+- `NUM_REQUIRED_ORACLES`: the number of oracle observations that must post a response to the chain
+- `COOLDOWN_BLOCKS`: the number of blocks that must elapse before the domain is allowed to move from the pending set to the active set.
+- `TIMEOUT_BLOCKS`: if insufficient oracle observations arrive within this number of blocks, the enrollment request expires.
+
+### `EnrollmentRequest`
+
+Invariants:
+- Non-concurrent: There cannot be multiple enrollment requests for a domain.
+
+The enrollment process involves:
+1. User publishes the domain's enrollment policy (see `server.md`) at:
+
+```
+https://<domain>/.well-known/webcat/enrollment.json
+```
+
+2. User submits an `EnrollmentRequest` to the chain.
+3. Transaction validation (`CheckTx`/`DeliverTx`)
+
+To unenroll, clients must remove the enrollment JSON and oracles verify that path produces a 404 or 410.
+
+#### Structure
+
+An `EnrollmentRequest` contains (TODO):
+- domain:
+- subdomains:
+- `signers`: a list of Ed25519 public keys
+
+#### `CheckTx`/`DeliverTx`
+
+We validate:
+- The maximum number of subdomains per domain has not been reached.
+
+The pending set tracks domains awaiting oracle observations. Each entry contains:
+
+- `domain`: the domain name
+- `request_height`: the block height when the request was submitted
+- `enrollment_data`: the signing data from the enrollment request
+- `oracle_observations`: oracle responses received so far (TODO do we need this?)
+
+Upon a new validated enrollment request:
+- If no pending request exists for the domain, we add it to the pending set.
+- If a pending request already exists for that domain, we remove the existing pending entry, add a new entry with the current block height as `request_height` with `oracle_observations` set to empty.
```

Comment:
If I understand this correctly, isn't there an attack here where I can keep submitting enrollment requests for a domain that I do not own, and keep restarting the process so that it never finalizes?

Reply:
we should probably just drop new requests for that site until the pending request is resolved - it either succeeds or `TIMEOUT_BLOCKS` pass and the request expires